imx_pwm_led.c

```diff
@@ -0,0 +1,2069 @@
+/* clang-format on */
+
+/** Helper macro to get the 'dev' and 'self' from a 'filp'.
+ *  miscdevice sets filp's private_data pointer to itself */
+#define DEV_SELF_FROM_FILP(filp)                                   \
+	struct device *dev =                                       \
+		((struct miscdevice *)filp->private_data)->parent; \
+	struct imx_pwm_led *self = dev_get_drvdata(dev)
+
+/*******************************************************************************
+ * TYPES
+ ******************************************************************************/
+
+/* clang-format off */
+/** Open object enum */
+enum open_object {
+	OPEN_OBJECT_NONE,
+	OPEN_OBJECT_FADE,
+	OPEN_OBJECT_CUE
+};
+
+/** Fade direction enum */
+enum fade_dir {
+	FADE_DIR_UNKNOWN,
+	FADE_DIR_INCREASING,
+	FADE_DIR_DECREASING,
+	FADE_DIR_NON_MONOTONIC
+};
+
+/** Permission mode enum */
+enum perm_mode {
+	PERM_MODE_NO_OWNER,
+	PERM_MODE_PID_OWNER,
+	PERM_MODE_NUM
+};
+/* clang-format on */
+
+/** LED device data struct */
+struct imx_pwm_led {
+	/** Pointer to parent device */
+	struct device *dev;
+	/** Lock to guard access */
+	struct mutex lock;
+	/** Peripheral clock */
+	struct clk *clk_per;
+	/** IPG clock */
+	struct clk *clk_ipg;
+	/** Base address of PWM registers */
+	void __iomem *mmio_base;
+	/** Physical address of PWM register base */
+	u32 mmio_base_phys;
+	/** Userspace interface */
+	struct miscdevice led_dev;
+	/** SDMA channel number */
+	u32 sdma_ch_num;
+	/** Pointer to the SDMA channel used by the driver */
+	struct sdma_channel *sdmac;
+	/** Device index */
+	uint led_idx;
+	/** Device name */
+	char *dev_name;
+	/** PID of owner */
+	pid_t owner_pid;
+	/** Currently open object */
+	enum open_object open_object;
+	/** Index of currently open object */
+	uint open_object_idx;
+	/** True if writing is ongoing */
+	int writing;
+	/** True if adaptive mode is enabled */
+	int adaptive;
+	/** Index of the currently playing fade, or -1 if not playing one */
+	int playing_fade_idx;
+	/** PWM FIFO sample period in microseconds */
+	uint pwm_fifo_period_us;
+};
+
+/** Fade struct */
+struct led_fade {
+	/** Buffer of u16 samples */
+	void *buf;
+	/** Physical address of buffer for the SDMA */
+	dma_addr_t buf_phys;
+	/** Length of fade data in **bytes** */
+	uint len;
+	/** Lock to guard access */
+	struct mutex lock;
+	/** PID of owner */
+	pid_t owner_pid;
+	/** Direction */
+	enum fade_dir dir;
+	/** Atomic reference counter for LEDs playing the fade */
+	atomic_t player_count;
+};
+
+/** Cue action struct */
+struct led_cue_action {
+	/** Index of LED */
+	u8 led_idx;
+	/** Type, see PWM_LED_CUE_ACTION_TYPE_* */
+	u8 type;
+	/** Value */
+	u16 val;
+};
+
+/** Cue struct */
+struct led_cue {
+	/** Buffer of actions */
+	struct led_cue_action *actions;
+	/** Length of action data in **bytes** */
+	uint len;
+	/** Lock to guard access */
+	struct mutex lock;
+	/** PID of owner */
+	pid_t owner_pid;
+};
+
+/** Module global data struct */
+struct glb_data {
```

Comment:
Agreed, I'll rename it as `struct module_data g_data`. `module_data` indicates that data is used by all devices, and `g_data` is nice and short, since this is used in a lot of places.

Reply:
Fixed